File: ios/Packages/HapiKit/Sources/HapiProtocol/Chat/NormalizeUser.swift
import Foundation

// Port of web/src/chat/normalizeUser.ts.

/// Port of `parseAttachments`: entries are accepted only when `id`,
/// `filename`, `mimeType` (strings), `size` (number) and `path` (string) are
/// all present; invalid entries are skipped and an empty result means
/// "no attachments" (`nil`).
func parseAttachments(_ raw: JSONValue?) -> [AttachmentMetadata]? {
    guard let items = raw?.arrayValue else { return nil }
    var attachments: [AttachmentMetadata] = []
    for item in items {
        guard let object = item.objectValue,
              let id = object["id"]?.stringValue,
              let filename = object["filename"]?.stringValue,
              let mimeType = object["mimeType"]?.stringValue,
              let size = object["size"]?.numberValue,
              let path = object["path"]?.stringValue,
              // The wire model stores byte sizes as Int; real sizes are
              // always integral (TS would accept a fractional number).
              let intSize = Int(exactly: size)
        else { continue }
        attachments.append(AttachmentMetadata(
            id: id,
            filename: filename,
            mimeType: mimeType,
            size: intSize,
            path: path,
            previewUrl: object["previewUrl"]?.stringValue
        ))
    }
    return attachments.isEmpty ? nil : attachments
}

/// Port of `normalizeUserRecord`.
func normalizeUserRecord(
    messageId: String,
    localId: String?,
    createdAt: Int,
    content: JSONValue,
    meta: JSONValue?
) -> NormalizedMessage? {
    if let text = content.stringValue {
        return NormalizedMessage(
            id: messageId,
            localId: localId,
            createdAt: createdAt,
            content: .user(text: text, attachments: nil),
            isSidechain: false,
            meta: meta
        )
    }

    if let object = content.objectValue,
       object["type"] == .string("text"),
       let text = object["text"]?.stringValue {
        let attachments = parseAttachments(object["attachments"])
        return NormalizedMessage(
            id: messageId,
            localId: localId,
            createdAt: createdAt,
            content: .user(text: text, attachments: attachments),
            isSidechain: false,
            meta: meta
        )
    }

    return nil
}

File: android/core/protocol/src/main/kotlin/app/hapi/protocol/chat/NormalizeUser.kt
package app.hapi.protocol.chat

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive

/** Port of `web/src/chat/normalizeUser.ts`. */

private fun parseAttachments(raw: JsonElement?): List<ChatAttachment>? {
    val array = raw as? JsonArray ?: return null
    val attachments = mutableListOf<ChatAttachment>()
    for (item in array) {
        val record = asObject(item) ?: continue
        val id = asString(record["id"]) ?: continue
        val filename = asString(record["filename"]) ?: continue
        val mimeType = asString(record["mimeType"]) ?: continue
        val size = asNumber(record["size"]) ?: continue
        val path = asString(record["path"]) ?: continue
        attachments.add(
            ChatAttachment(
                id = id,
                filename = filename,
                mimeType = mimeType,
                size = size,
                path = path,
                previewUrl = asString(record["previewUrl"]),
            )
        )
    }
    return attachments.takeIf { it.isNotEmpty() }
}

fun normalizeUserRecord(
    messageId: String,
    localId: String?,
    createdAt: Long,
    content: JsonElement?,
    meta: JsonElement?,
): NormalizedMessage? {
    if (content is JsonPrimitive && content.isString) {
        return NormalizedMessage.User(
            id = messageId,
            localId = localId,
            createdAt = createdAt,
            text = content.content,
            isSidechain = false,
            meta = meta,
        )
    }

    val record = asObject(content)
    if (record != null && asString(record["type"]) == "text") {
        val text = asString(record["text"])
        if (text != null) {
            return NormalizedMessage.User(
                id = messageId,
                localId = localId,
                createdAt = createdAt,
                text = text,
                attachments = parseAttachments(record["attachments"]),
                isSidechain = false,
                meta = meta,
            )
        }
    }

    return null
}
